<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/hpdp/yumm/Assignment6-Yumm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Mastering Big Data Handling - Assignment 6

## Group Members

|    Name           |           Matric Number       |                      
|-------------------|-------------------------------|
|   Nur Khuzairie Zakwan Bin Mohd Zamri| A21EC0112     |
|Muhammad Farhan Bin Ibrahim|        A21EC0072      |

---
## About Dataset

The data in this dataset was collected by Properati.

**Context**

One of the best applications of data science and machine learning in general is the real estate business. This data set provides data for those who want to make data analysis and use of machine learning models to perform multiple tasks and generate new insights.

**Content**

It consists of a .csv where each row contains a publication.
The .csv contains no missing data, this means that it is almost ready for use and model training. The only thing necessary is to convert the "string" type data into numerical data.

**Columns**

id - Notice identifier. It is not unique: if the notification is updated by the real estate agency (new version of the notification) a new record is created with the same id but different dates: registration and cancellation.

operation_type - Type of operation (these are all sales, can be removed).

l2 - Administrative level 2: usually province

l3 - Administrative level 3: usually city

lat - Latitude.

lon - Longitude.

price - Price published in the ad.

property_type - Type of property (House, Apartment, PH).

rooms - Number of rooms (useful in Argentina).

bathrooms - Number of bathrooms.

start_date - Date when the ad was created.

end_date - Date of termination of the advertisement.

created_on - Date when the first version of the notice was created.

surface_total - Total area in m².

surface_covered - Covered area in m².

title - Title of the advertisement.

description - Description of the advertisement.

ad_type - Type of ad (Property, Development/Project).

**Acknowledgements**

The data in this dataset was collected by Properati.

# Strategies for Big Datasets: Apply five smart strategies to handle large datasets effectively:

Loading the Dataset:

In [1]:
!mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json
! kaggle datasets download -d msorondo/argentina-venta-de-propiedades
!unzip argentina-venta-de-propiedades.zip

 98% 692M/703M [00:06<00:00, 137MB/s]
100% 703M/703M [00:06<00:00, 114MB/s]
Archive:  argentina-venta-de-propiedades.zip
  inflating: ar_properties.csv       
  inflating: ar_properties_crude.csv  
  inflating: co_properties_crude.csv  
  inflating: ec_properties_crude.csv  
  inflating: pe_properties_crude.csv  
  inflating: uy_properties_crude.csv  


Settings up the environment

Import necessaries tools and libraries

In [6]:
import pandas as pd
import numpy as np
import os
import time
import datetime
!pip install memory_profiler

Remove unused datasets and folder to minimize disk storage

In [7]:
os.remove('ar_properties_crude.csv')
os.remove('co_properties_crude.csv')
os.remove('ec_properties_crude.csv')
os.remove('pe_properties_crude.csv')
os.remove('uy_properties_crude.csv')
os.remove('argentina-venta-de-propiedades.zip')

FileNotFoundError: ignored

Read 'ar_properties.csv' dataset.

In [8]:
start_time = time.time()
df = pd.read_csv('ar_properties.csv')
end_time = time.time()

time_taken = end_time - start_time
print(f"Time taken to read the CSV file: {time_taken} seconds")

Time taken to read the CSV file: 23.549723148345947 seconds


Overview of the dataset

In [9]:
df.head(5)

id    ad_type  start_date    end_date  created_on  \
0  wdQ5hWhv8P14T7Sh9g4QCg==  Propiedad  2020-12-25  9999-12-31  2020-12-25   
1  nnMBYZ4RMRY+vm753EtA+g==  Propiedad  2020-12-25  9999-12-31  2020-12-25   
2  +dnVA1K6JxzL1zAjOEQ1pA==  Propiedad  2020-12-25  2020-12-29  2020-12-25   
3  dLHXKN5/sRZpm9Yk0yI2nA==  Propiedad  2020-12-25  2020-12-29  2020-12-25   
4  wtw/k887EPipd37UYHKb1Q==  Propiedad  2020-12-25  9999-12-31  2020-12-25   

         lat        lon         l1                        l2        l3  ...  \
0 -32.716652 -68.642692  Argentina                   Mendoza       NaN  ...   
1 -24.797723 -65.467514  Argentina                     Salta       NaN  ...   
2 -34.919373 -58.020591  Argentina    Bs.As. G.B.A. Zona Sur  La Plata  ...   
3 -34.919455 -58.024807  Argentina    Bs.As. G.B.A. Zona Sur  La Plata  ...   
4 -34.364924 -58.783143  Argentina  Bs.As. G.B.A. Zona Norte   Escobar  ...   

  bathrooms surface_total  surface_covered  price  currency  price_period  \
0       NaN         350.0            350.0    NaN       NaN           NaN   
1       NaN        1541.0           1541.0    NaN       NaN       Mensual   
2       NaN        1000.0           1000.0    NaN       NaN       Mensual   
3       NaN        1000.0           1000.0    NaN       NaN       Mensual   
4       NaN       18164.0          18164.0    NaN       NaN       Mensual   

                                               title  \
0                     Excelentes Lotes Sobre Ruta 34   
1  TERRENO + VENTA + JARDINES DE SAN LORENZO +150...   
2               Lote en Venta de 1000 m2 en La Plata   
3               Lote en Venta de 1000 m2 en La Plata   
4                                 PANAMERICANA 47300   

                                         description  property_type  \
0  Corredor Responsable: VICTOR E. MONTIVERO - C....           Lote   
1  Corredor Responsable: Pablo Castañeda - C.U.C....           Lote   
2  Corredor Responsable: Rico Sebastián - Martill...           Lote   
3  Corredor Responsable: Rico Sebastián - Martill...           Lote   
4  Nave principal 66 x 90 m:    6005 m2 cubiertos...           Otro   

  operation_type  
0          Venta  
1          Venta  
2          Venta  
3          Venta  
4          Venta  

[5 rows x 25 columns]

In [10]:
df.shape

(1000000, 25)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  object 
 1   ad_type          1000000 non-null  object 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              849189 non-null   float64
 6   lon              848255 non-null   float64
 7   l1               1000000 non-null  object 
 8   l2               1000000 non-null  object 
 9   l3               946673 non-null   object 
 10  l4               239496 non-null   object 
 11  l5               5001 non-null     object 
 12  l6               0 non-null        float64
 13  rooms            526423 non-null   float64
 14  bedrooms         466213 non-null   float64
 15  bathrooms        776733 non-null   float64
 16  surface_total    37

### Load Less Data:
Strategically load only the essential portions of the dataset to optimize memory usage.


Drop columns that are unused

In [12]:
columns_to_drop = ['l1', 'l2', 'l3', 'l4', 'l5', 'l6', 'rooms', 'bedrooms', 'bathrooms']

start_time = time.time()
df_less1 = pd.read_csv('ar_properties.csv', usecols=lambda column: column not in columns_to_drop)
end_time = time.time()

time_taken = end_time - start_time
print(f"Time taken to read the CSV file: {time_taken} seconds")

Time taken to read the CSV file: 17.854825258255005 seconds


In [13]:
df_less1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 16 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  object 
 1   ad_type          1000000 non-null  object 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              849189 non-null   float64
 6   lon              848255 non-null   float64
 7   surface_total    379501 non-null   float64
 8   surface_covered  381451 non-null   float64
 9   price            963098 non-null   float64
 10  currency         961719 non-null   object 
 11  price_period     363015 non-null   object 
 12  title            1000000 non-null  object 
 13  description      999982 non-null   object 
 14  property_type    1000000 non-null  object 
 15  operation_type   1000000 non-null  object 
dtypes: float64(5), obje

In [14]:
print('Missing values before: ')
df_less1.isna().sum()

Missing values before: 


id                      0
ad_type                 0
start_date              0
end_date                0
created_on              0
lat                150811
lon                151745
surface_total      620499
surface_covered    618549
price               36902
currency            38281
price_period       636985
title                   0
description            18
property_type           0
operation_type          0
dtype: int64

Drop missing values in each column

In [15]:
df_less1.dropna(subset=['lat'], inplace=True)
df_less1.dropna(subset=['lon'], inplace=True)
df_less1.dropna(subset=['surface_total'], inplace=True)
df_less1.dropna(subset=['surface_covered'], inplace=True)
df_less1.dropna(subset=['price'], inplace=True)
df_less1.dropna(subset=['currency'], inplace=True)
df_less1.dropna(subset=['price_period'], inplace=True)
df_less1.dropna(subset=['description'], inplace=True)

As we can see, the memory usage reduce from 190.7 MB (**df**) to 17.5 MB (**df_less1**)

In [16]:
df_less1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 134863 entries, 31 to 999999
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   id               134863 non-null  object 
 1   ad_type          134863 non-null  object 
 2   start_date       134863 non-null  object 
 3   end_date         134863 non-null  object 
 4   created_on       134863 non-null  object 
 5   lat              134863 non-null  float64
 6   lon              134863 non-null  float64
 7   surface_total    134863 non-null  float64
 8   surface_covered  134863 non-null  float64
 9   price            134863 non-null  float64
 10  currency         134863 non-null  object 
 11  price_period     134863 non-null  object 
 12  title            134863 non-null  object 
 13  description      134863 non-null  object 
 14  property_type    134863 non-null  object 
 15  operation_type   134863 non-null  object 
dtypes: float64(5), object(11)
memory usag

In [17]:
print('Missing values after: ')
df_less1.isna().sum()

Missing values after: 


id                 0
ad_type            0
start_date         0
end_date           0
created_on         0
lat                0
lon                0
surface_total      0
surface_covered    0
price              0
currency           0
price_period       0
title              0
description        0
property_type      0
operation_type     0
dtype: int64

### Use Chunking:
Process the data in smaller pieces to avoid memory issues.


In [18]:
start_time = time.time()

# Reading the data in chunks
data = pd.read_csv('ar_properties.csv', chunksize=300000)
dfs = []
for chunk in data:
    dfs.append(chunk)

# Concatenating the chunks together
df_less2 = pd.concat(dfs)

end_time = time.time()

time_taken = end_time - start_time
print(f"Time taken to read and concatenate the CSV file in chunks: {time_taken} seconds")


Time taken to read and concatenate the CSV file in chunks: 18.3456974029541 seconds


In [19]:
df_less2.head(5)

id    ad_type  start_date    end_date  created_on  \
0  wdQ5hWhv8P14T7Sh9g4QCg==  Propiedad  2020-12-25  9999-12-31  2020-12-25   
1  nnMBYZ4RMRY+vm753EtA+g==  Propiedad  2020-12-25  9999-12-31  2020-12-25   
2  +dnVA1K6JxzL1zAjOEQ1pA==  Propiedad  2020-12-25  2020-12-29  2020-12-25   
3  dLHXKN5/sRZpm9Yk0yI2nA==  Propiedad  2020-12-25  2020-12-29  2020-12-25   
4  wtw/k887EPipd37UYHKb1Q==  Propiedad  2020-12-25  9999-12-31  2020-12-25   

         lat        lon         l1                        l2        l3  ...  \
0 -32.716652 -68.642692  Argentina                   Mendoza       NaN  ...   
1 -24.797723 -65.467514  Argentina                     Salta       NaN  ...   
2 -34.919373 -58.020591  Argentina    Bs.As. G.B.A. Zona Sur  La Plata  ...   
3 -34.919455 -58.024807  Argentina    Bs.As. G.B.A. Zona Sur  La Plata  ...   
4 -34.364924 -58.783143  Argentina  Bs.As. G.B.A. Zona Norte   Escobar  ...   

  bathrooms surface_total  surface_covered  price  currency  price_period  \
0       NaN         350.0            350.0    NaN       NaN           NaN   
1       NaN        1541.0           1541.0    NaN       NaN       Mensual   
2       NaN        1000.0           1000.0    NaN       NaN       Mensual   
3       NaN        1000.0           1000.0    NaN       NaN       Mensual   
4       NaN       18164.0          18164.0    NaN       NaN       Mensual   

                                               title  \
0                     Excelentes Lotes Sobre Ruta 34   
1  TERRENO + VENTA + JARDINES DE SAN LORENZO +150...   
2               Lote en Venta de 1000 m2 en La Plata   
3               Lote en Venta de 1000 m2 en La Plata   
4                                 PANAMERICANA 47300   

                                         description  property_type  \
0  Corredor Responsable: VICTOR E. MONTIVERO - C....           Lote   
1  Corredor Responsable: Pablo Castañeda - C.U.C....           Lote   
2  Corredor Responsable: Rico Sebastián - Martill...           Lote   
3  Corredor Responsable: Rico Sebastián - Martill...           Lote   
4  Nave principal 66 x 90 m:    6005 m2 cubiertos...           Otro   

  operation_type  
0          Venta  
1          Venta  
2          Venta  
3          Venta  
4          Venta  

[5 rows x 25 columns]

In [20]:
print(df_less2.shape)

(1000000, 25)


In [21]:
df_less2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  object 
 1   ad_type          1000000 non-null  object 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              849189 non-null   float64
 6   lon              848255 non-null   float64
 7   l1               1000000 non-null  object 
 8   l2               1000000 non-null  object 
 9   l3               946673 non-null   object 
 10  l4               239496 non-null   object 
 11  l5               5001 non-null     object 
 12  l6               0 non-null        float64
 13  rooms            526423 non-null   float64
 14  bedrooms         466213 non-null   float64
 15  bathrooms        776733 non-null   float64
 16  surface_total    37

### Optimize Data Types:
Fine-tune data types to maximize efficiency and minimize memory consumption.


In [22]:
df.dtypes

id                  object
ad_type             object
start_date          object
end_date            object
created_on          object
lat                float64
lon                float64
l1                  object
l2                  object
l3                  object
l4                  object
l5                  object
l6                 float64
rooms              float64
bedrooms           float64
bathrooms          float64
surface_total      float64
surface_covered    float64
price              float64
currency            object
price_period        object
title               object
description         object
property_type       object
operation_type      object
dtype: object

In [23]:
df_opt = df.copy()

In [24]:
df_opt.dtypes

id                  object
ad_type             object
start_date          object
end_date            object
created_on          object
lat                float64
lon                float64
l1                  object
l2                  object
l3                  object
l4                  object
l5                  object
l6                 float64
rooms              float64
bedrooms           float64
bathrooms          float64
surface_total      float64
surface_covered    float64
price              float64
currency            object
price_period        object
title               object
description         object
property_type       object
operation_type      object
dtype: object

In [25]:
df_opt.head(5)

id    ad_type  start_date    end_date  created_on  \
0  wdQ5hWhv8P14T7Sh9g4QCg==  Propiedad  2020-12-25  9999-12-31  2020-12-25   
1  nnMBYZ4RMRY+vm753EtA+g==  Propiedad  2020-12-25  9999-12-31  2020-12-25   
2  +dnVA1K6JxzL1zAjOEQ1pA==  Propiedad  2020-12-25  2020-12-29  2020-12-25   
3  dLHXKN5/sRZpm9Yk0yI2nA==  Propiedad  2020-12-25  2020-12-29  2020-12-25   
4  wtw/k887EPipd37UYHKb1Q==  Propiedad  2020-12-25  9999-12-31  2020-12-25   

         lat        lon         l1                        l2        l3  ...  \
0 -32.716652 -68.642692  Argentina                   Mendoza       NaN  ...   
1 -24.797723 -65.467514  Argentina                     Salta       NaN  ...   
2 -34.919373 -58.020591  Argentina    Bs.As. G.B.A. Zona Sur  La Plata  ...   
3 -34.919455 -58.024807  Argentina    Bs.As. G.B.A. Zona Sur  La Plata  ...   
4 -34.364924 -58.783143  Argentina  Bs.As. G.B.A. Zona Norte   Escobar  ...   

  bathrooms surface_total  surface_covered  price  currency  price_period  \
0       NaN         350.0            350.0    NaN       NaN           NaN   
1       NaN        1541.0           1541.0    NaN       NaN       Mensual   
2       NaN        1000.0           1000.0    NaN       NaN       Mensual   
3       NaN        1000.0           1000.0    NaN       NaN       Mensual   
4       NaN       18164.0          18164.0    NaN       NaN       Mensual   

                                               title  \
0                     Excelentes Lotes Sobre Ruta 34   
1  TERRENO + VENTA + JARDINES DE SAN LORENZO +150...   
2               Lote en Venta de 1000 m2 en La Plata   
3               Lote en Venta de 1000 m2 en La Plata   
4                                 PANAMERICANA 47300   

                                         description  property_type  \
0  Corredor Responsable: VICTOR E. MONTIVERO - C....           Lote   
1  Corredor Responsable: Pablo Castañeda - C.U.C....           Lote   
2  Corredor Responsable: Rico Sebastián - Martill...           Lote   
3  Corredor Responsable: Rico Sebastián - Martill...           Lote   
4  Nave principal 66 x 90 m:    6005 m2 cubiertos...           Otro   

  operation_type  
0          Venta  
1          Venta  
2          Venta  
3          Venta  
4          Venta  

[5 rows x 25 columns]

In [26]:
df_opt['id'] = df_opt['id'].astype('string')
df_opt['ad_type'] = df_opt['ad_type'].astype('string')
df_opt['currency'] = df_opt['currency'].astype('string')
df_opt['price_period'] = df_opt['price_period'].astype('string')
df_opt['title'] = df_opt['title'].astype('string')
df_opt['description'] = df_opt['description'].astype('string')
df_opt['property_type'] = df_opt['property_type'].astype('string')
df_opt['operation_type'] = df_opt['operation_type'].astype('string')

In [27]:
df_opt['lat'] = df_opt['lat'].astype('float16')
df_opt['lon'] = df_opt['lon'].astype('float16')
df_opt['surface_total'] = df_opt['surface_total'].astype('float16')
df_opt['surface_covered'] = df_opt['surface_covered'].astype('float16')
df_opt['price'] = df_opt['price'].astype('float16')

By changing data types of each column also help to reduce the memory usage in this dataset from 190.7+ MB (**df**) to 162.1 MB (**df_opt**). If you want to reduce it more, you can combine this method with load less data method.

In [28]:
df_opt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  string 
 1   ad_type          1000000 non-null  string 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              849189 non-null   float16
 6   lon              848255 non-null   float16
 7   l1               1000000 non-null  object 
 8   l2               1000000 non-null  object 
 9   l3               946673 non-null   object 
 10  l4               239496 non-null   object 
 11  l5               5001 non-null     object 
 12  l6               0 non-null        float64
 13  rooms            526423 non-null   float64
 14  bedrooms         466213 non-null   float64
 15  bathrooms        776733 non-null   float64
 16  surface_total    37

### Sampling :
Implement sampling methodologies to extract meaningful insights from a subset of the dataset.


##Random Sampling:
- Use only 100000 rows from the entire rows
- Equal probability of the selection by using random sampling


In [29]:
file_path='/content/ar_properties.csv'
sample_size = 100000

# Random Sample

df_random_sample = pd.read_csv(file_path).sample(n=sample_size, random_state=42)
print("Random Sample:")
print(df_random_sample.head())

# Stratified Sample
df_stratified_sample = pd.read_csv(file_path)
stratified_sample = df_stratified_sample.groupby('ad_type', group_keys=False).apply(lambda x: x.sample(min(len(x), sample_size), random_state=42))
print("\nStratified Sample:")
print(stratified_sample.head())

# Time-Based Sample
df_time_based_sample = pd.read_csv(file_path)
df['created_on'] = pd.to_datetime(df['created_on'])
df_time_based_sample = df_time_based_sample.sort_values(by='created_on').head(sample_size)
print("\nTime-Based Sample:")
print(df_time_based_sample.head())

# Percentage-Based Sample
df_percentage_sample = pd.read_csv(file_path).sample(frac=0.1, random_state=42)
print("\nPercentage-Based Sample:")
print(df_percentage_sample.head())


Random Sample:
                              id    ad_type  start_date    end_date  \
987231  yLn9byy3iVivVO9TQHPztg==  Propiedad  2020-02-11  2020-02-13   
79954   UzJHl4kI7+/Gf8U9niRFrw==  Propiedad  2020-08-08  2020-10-04   
567130  pbZUQph+oNbxkq3t6vsWhg==  Propiedad  2020-10-16  9999-12-31   
500891  1CAUxm8sj/60MLUx/E4PMg==  Propiedad  2020-02-14  2020-12-03   
55399   qm3O2Eg5bqoxoRL4pUwuSQ==  Propiedad  2021-01-11  9999-12-31   

        created_on        lat        lon         l1                        l2  \
987231  2020-02-11 -34.507628 -58.525211  Argentina  Bs.As. G.B.A. Zona Norte   
79954   2020-08-08 -29.164336 -59.671764  Argentina                  Santa Fe   
567130  2020-10-16 -34.927879 -54.935719    Uruguay                 Maldonado   
500891  2020-02-14 -34.658158 -58.822441  Argentina  Bs.As. G.B.A. Zona Oeste   
55399   2021-01-11 -34.602601 -58.542348  Argentina  Bs.As. G.B.A. Zona Oeste   

                     l3  ... bathrooms surface_total  surface_covered  

In [30]:
import psutil
import time
import os
import sys

memory_before_random = psutil.Process().memory_info().rss / 1024 / 1024
start_time_random = time.time()
df_random_sample = pd.read_csv(file_path).sample(n=sample_size, random_state=42)
end_time_random = time.time()
memory_after_random = psutil.Process().memory_info().rss / 1024 / 1024
time_taken_random = end_time_random - start_time_random
file_size_random = os.path.getsize(file_path) / (1024 * 1024)

print("Random Sample:")
print(df_random_sample.head())
print(f"Memory used: {memory_after_random - memory_before_random} MB")
print(f"Time taken: {time_taken_random} seconds")
print(f"File size: {file_size_random} MB")
print("\n---")

Random Sample:
                              id    ad_type  start_date    end_date  \
987231  yLn9byy3iVivVO9TQHPztg==  Propiedad  2020-02-11  2020-02-13   
79954   UzJHl4kI7+/Gf8U9niRFrw==  Propiedad  2020-08-08  2020-10-04   
567130  pbZUQph+oNbxkq3t6vsWhg==  Propiedad  2020-10-16  9999-12-31   
500891  1CAUxm8sj/60MLUx/E4PMg==  Propiedad  2020-02-14  2020-12-03   
55399   qm3O2Eg5bqoxoRL4pUwuSQ==  Propiedad  2021-01-11  9999-12-31   

        created_on        lat        lon         l1                        l2  \
987231  2020-02-11 -34.507628 -58.525211  Argentina  Bs.As. G.B.A. Zona Norte   
79954   2020-08-08 -29.164336 -59.671764  Argentina                  Santa Fe   
567130  2020-10-16 -34.927879 -54.935719    Uruguay                 Maldonado   
500891  2020-02-14 -34.658158 -58.822441  Argentina  Bs.As. G.B.A. Zona Oeste   
55399   2021-01-11 -34.602601 -58.542348  Argentina  Bs.As. G.B.A. Zona Oeste   

                     l3  ... bathrooms surface_total  surface_covered  

In [31]:
memory_before_percentage = psutil.Process().memory_info().rss / 1024 / 1024
start_time_percentage = time.time()
df_percentage_sample = pd.read_csv(file_path).sample(frac=0.1, random_state=42)
end_time_percentage = time.time()
memory_after_percentage = psutil.Process().memory_info().rss / 1024 / 1024
time_taken_percentage = end_time_percentage - start_time_percentage
file_size_percentage = os.path.getsize(file_path) / (1024 * 1024)

print("Percentage-Based Sample:")
print(df_percentage_sample.head())
print(f"Memory used: {memory_after_percentage - memory_before_percentage} MB")
print(f"Time taken: {time_taken_percentage} seconds")
print(f"File size: {file_size_percentage} MB")

Percentage-Based Sample:
                              id    ad_type  start_date    end_date  \
987231  yLn9byy3iVivVO9TQHPztg==  Propiedad  2020-02-11  2020-02-13   
79954   UzJHl4kI7+/Gf8U9niRFrw==  Propiedad  2020-08-08  2020-10-04   
567130  pbZUQph+oNbxkq3t6vsWhg==  Propiedad  2020-10-16  9999-12-31   
500891  1CAUxm8sj/60MLUx/E4PMg==  Propiedad  2020-02-14  2020-12-03   
55399   qm3O2Eg5bqoxoRL4pUwuSQ==  Propiedad  2021-01-11  9999-12-31   

        created_on        lat        lon         l1                        l2  \
987231  2020-02-11 -34.507628 -58.525211  Argentina  Bs.As. G.B.A. Zona Norte   
79954   2020-08-08 -29.164336 -59.671764  Argentina                  Santa Fe   
567130  2020-10-16 -34.927879 -54.935719    Uruguay                 Maldonado   
500891  2020-02-14 -34.658158 -58.822441  Argentina  Bs.As. G.B.A. Zona Oeste   
55399   2021-01-11 -34.602601 -58.542348  Argentina  Bs.As. G.B.A. Zona Oeste   

                     l3  ... bathrooms surface_total  surface

### Parallelize with Dask:
- Dask is a powerful library that extends pandas to enable parallel and distributed computing. It's particularly useful for handling larger-than-memory datasets.
- Scalability: Dask allows you to scale your computation from a single machine to a cluster of machines. This makes it suitable for handling large datasets and computationally intensive tasks that may not fit into the memory of a single machine.

- Parallel Processing: Dask enables parallel processing by breaking down a task into smaller tasks and executing them concurrently. This can significantly speed up the computation time, especially for tasks that can be parallelized.

- Lazy Evaluation: Dask uses lazy evaluation, meaning that it builds up a task graph without immediately executing the operations. This allows for more efficient task scheduling and resource allocation. The actual computation is triggered only when the results are explicitly requested, allowing for optimization opportunities.

In [32]:
pip install dask

In [33]:
import dask.dataframe as dd


In [34]:
file_path = '/content/ar_properties.csv'

ddf = dd.read_csv(file_path, engine="python", on_bad_lines="skip")


file_size_dask = os.path.getsize(file_path) / (1024 * 1024)

memory_before_dask = psutil.Process().memory_info().rss / 1024 / 1024
start_time_dask = time.time()

end_time_dask = time.time()
memory_after_dask = psutil.Process().memory_info().rss / 1024 / 1024
time_taken_dask = end_time_dask - start_time_dask

print("Dask Processing:")
print(f"Memory used: {memory_after_dask - memory_before_dask} MB")
print(f"Time taken: {time_taken_dask} seconds")
print(f"File size: {file_size_dask} MB")

Dask Processing:
Memory used: 0.0 MB
Time taken: 8.940696716308594e-05 seconds
File size: 908.2618713378906 MB


In [35]:
num_rows_before_compute = len(ddf)

num_rows_after_compute = len(ddf)
print("Dask Processing:")
print(f"Number of Rows Before Computation: {num_rows_before_compute}")
print(f"Number of Rows After Computation: {num_rows_after_compute}")

Dask Processing:
Number of Rows Before Computation: 1000023
Number of Rows After Computation: 1000023


## Comparative Analysis


| Strategy                | Memory Used (MB) | Time Taken (seconds) | File Size (MB) |
|-------------------------|-------------------|-----------------------|----------------|
| Load Less Data          | 17.5             | 17.85                 | 908.26         |
| Chunking                | 190.7             | 18.35                 | -              |
| Data Type Optimization  |                   |                       |                |
| - Optimized Data Types  | 162.1             | -                 | -              |
| Sampling (Random)       | 130.80              | 18.99                 | 908.26         |
| Sampling (Percentage)   | 0.33           | 20.31                 | 908.26         |
| Parallelize with Dask   | 0.0               | 8.94                | 908.26         |



##Conclusion

After analyzing various data processing strategies, it is evident that each approach has its own strengths and trade-offs.

- **Load Less Data:** This strategy demonstrates moderate memory usage (17.5 MB) and reasonable processing time (17.85 seconds). The resulting file size is consistent at 908.26 MB.

- **Chunking:** While this strategy incurs high memory usage (190.7 MB), it compensates with a relatively quick processing time of 18.35 seconds. However, specific details about the file size are not provided.

- **Data Type Optimization:** The optimization of data types significantly reduces memory usage to 162.1 MB, but the processing time and file size are not specified.

- **Sampling (Random):** This strategy involves substantial memory usage (130.80 MB) and a processing time of 18.99 seconds. The resulting file size remains constant at 908.26 MB.

- **Sampling (Percentage):** With minimal memory usage (0.33 MB), this strategy takes a slightly longer processing time of 20.31 seconds. The resulting file size remains consistent at 908.26 MB.

- **Parallelize with Dask:** This strategy stands out with virtually zero memory usage (0.0 MB) and a notably fast processing time of 8.94 seconds. The resulting file size is 908.26 MB.

In conclusion, the choice of strategy depends on the specific requirements of the data processing task. If minimizing memory usage is critical, the "Parallelize with Dask" strategy proves to be highly efficient. However, other strategies may be more suitable depending on the trade-offs between memory usage, processing time, and specific processing needs.